maze_transformer/evaluation/path_evals.py

```diff
@@ -0,0 +1,109 @@
+
+
+def path_as_segments_iter(path: MazePath) -> Iterable[tuple]:
+    """
+    Iterate over the segments of a path.
+    """
+    i: int
+    n_s: Coord | CoordTup
+    n_e: Coord | CoordTup
+    for i, n_s in enumerate(path[:-1]):
+        n_e = path[i + 1]
+        yield (n_s, n_e)
+
+
+class PathEvals:
+    """array path based eval functions. first path is always the "ground truth" path"""
+
+    @staticmethod
+    def node_overlap(
+        maze: LatticeMaze, solution: MazePath, prediction: MazePath, /
+    ) -> float:
+        """number of shared nodes (any order) / total number of (unique) nodes"""
+        if len(prediction) <= 1:
+            return 0.0
+
+        n_shared: int = 0
+        solution_set = {tuple(coord) for coord in solution}
+        prediction_set = {tuple(coord) for coord in prediction}
+
+        for coord in solution:
+            if tuple(coord) in prediction_set:
+                n_shared += 1
+        return n_shared / len(solution_set)
+
+    @staticmethod
+    def num_connections_adjacent_lattice(
+        maze: LatticeMaze, solution: MazePath, prediction: MazePath, /
+    ) -> float:
+        """number of the connections in prediction which actually connect nodes that are adjacent on the lattice, ignoring if they are adjacent on the maze"""
+        if len(prediction) <= 1:
+            return 0.0
+
+        n_adj: int = 0
+        for n_s, n_e in path_as_segments_iter(prediction):
+            # print(f"{n_s = } {n_e = }")
+
+            if (np.abs(n_s - n_e).sum() == 1).all():
+                n_adj += 1
+
+        return n_adj
+
+    @staticmethod
+    def fraction_connections_adjacent_lattice(
+        maze: LatticeMaze, solution: MazePath, prediction: MazePath, /
+    ) -> float:
+        """fraction of the connections in prediction which actually connect nodes that are adjacent on the lattice, ignoring if they are adjacent on the maze"""
+
+        return PathEvals.num_connections_adjacent_lattice(
+            maze, solution, prediction
+        ) / len(prediction)
+
+    @staticmethod
+    def num_connections_adjacent(
+        maze: LatticeMaze, solution: MazePath, prediction: MazePath, /
+    ) -> float:
+        """number of connections in prediction which are are valid paths on the maze"""
+
+        if len(prediction) <= 1:
+            return 0.0
+
+        n_connected: int = 0
+        for n_s, n_e in path_as_segments_iter(prediction):
+            if maze.nodes_connected(n_s, n_e):
+                n_connected += 1
+
+        return n_connected
+
+    @staticmethod
+    def fraction_connections_adjacent(
+        maze: LatticeMaze, solution: MazePath, prediction: MazePath, /
+    ) -> float:
+        """fraction of connections in prediction which are are valid paths on the maze"""
+
+        return PathEvals.num_connections_adjacent(maze, solution, prediction) / len(
+            prediction
+        )
+
+    @classmethod
+    def all_functions(cls) -> dict[str, PathEvalFunction]:
```

Comment:
dumb question but what is this for?

Reply:
I'm fairly new to Python, so if you find anything on this I'd like to know too. 🙂

Reply:
the python docs are relatively unhelpful with this afaik, I found these two helpful SO threads:
- https://stackoverflow.com/questions/48029249/python-dict
- https://stackoverflow.com/questions/19907442/explain-dict-attribute

the tl;dr is that it contains all attributes of an object, but all the built-in ones like `__module__`, `__doc__`, and `__dict__` itself start with `__` to signify that they are "magic" methods. Since we only inherit from `object`, there shouldn't be anything other than a built-in in the class.

Starting with `_` is [convention in python](https://peps.python.org/pep-0008/#descriptive-naming-styles) for "this is a private method" or otherwise "don't use this unless you know what you're doing".

> Tangent: the PEP style guide says never to invent double-underscore attributes and only use built-ins, but I may be guilty of doing this sometimes 😢

Reply:
sorry if I'm stepping on your toes Dan, I took care of this in #[bf407c5](https://github.com/AISC-understanding-search/maze-transformer/pull/134/commits/bf407c599a330c87c577e133d84ddb3fb748a9bf)